### PR TITLE
data_structures/linked_list: Adding __str__() function #3960

### PR DESCRIPTION
**Pull request by arif599**
### **Describe your change:**
Added a __str__() function to the Linked List class that will print the contents in the Linked List with arrows that point to the next value

see: #3960 

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by arif599** [comment]
I'm not really sure what went wrong during the build check.

**Comment by xcodz-dot** [comment]
Your doctest failed, please verify it

File: data_structures/linked_list/__init__.py [hunk 1]
```diff
@@ -25,6 +25,26 @@ def remove(self):
...
+        if not self.is_empty:
+            return "Linked List is empty."
+        else:
+            iterate = self.head
+            iteam_str = ""
+            while iterate:
+                iteam_str += f"{iterate.item} --> "
+                iterate = iterate.next
+
+            return iteam_str
```
Comment (xcodz-dot) [inline data_structures/linked_list/__init__.py:46]:
```suggestion
            return iteam_str.strip()
```
The problem here is that the pre-commit just trims the whitespaces so the doctests fail

```python
>>> str(linked_list).strip() == str(linked_list)
False

>>> str(linked_list).strip() != str(linked_list)
True
```

**Review by xcodz-dot — CHANGES_REQUESTED** [review]

**Comment by arif599** [comment]
Thanks for bringing that up. I've never used doctests in python before so I'll do some more research. Based upon looking at other Pull Requests, I thought it was just a comment that shows others how to use it but I didn't know it also gets executed.

File: data_structures/linked_list/__init__.py [hunk 2]
```diff
@@ -25,6 +25,26 @@ def remove(self):
...
+
     def __len__(self):
```
Comment (xcodz-dot) [inline data_structures/linked_list/__init__.py:48]:
```suggestion
        >>> linked_list.add(9)
        >>> print(linked_list)
        9 --> 14 --> 23
        """
        if not self.is_empty:
            return "Linked List is empty."
        else:
            iterate = self.head
            item_str = ""
            item_list = []
            while iterate:
                item_list.append(iterate.item)
                iterate = iterate.next

            item_list = [str(item) for item in item_list]
            item_str = " --> ".join(item_list)

            return item_str

    def __len__(self):
```

**Review by xcodz-dot — CHANGES_REQUESTED** [review]
Will fix any doctest errors, make the output better to look at without any extra arrow at the end

**Comment by arif599** [comment]
> Will fix any doctest errors, make the output better to look at without any extra arrow at the end
> 
> 
> 
> also a spelling error `iteam` -> `item`

Thank you for catching that spelling mistake. I really like the format of the output, I committed the changes you added.

File: data_structures/linked_list/__init__.py [hunk 3]
```diff
@@ -25,6 +25,30 @@ def remove(self):
     def is_empty(self):
         return self.head is None
 
+    def __str__(self):
```
Comment (dhruvmanila) [inline data_structures/linked_list/__init__.py:28]:
```suggestion
    def __str__(self) -> str:
```
Please add a return type hint.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]
If possible, can you add type hints for other functions/parameters as well? Thanks if you do, otherwise after this small change the PR is good to go.

**Comment by arif599** [comment]
Yeah definitely, I'll include type hinting and also add comments to describe some of the functionality. It's midnight for me, so I will have it done by tomorrow morning.

**Comment by xcodz-dot** [comment]
Great, Also your doctests were failing because of the whitespace in that line which was required but trimmed by pre commit

File: data_structures/linked_list/__init__.py [hunk 4]
```diff
@@ -1,19 +1,28 @@
...
+Nodes contain data and also may link to other nodes:
+    - Head Node: First node, the address of the
+                 head node gives us access of the complete list
+    - Last node: points to null
+"""
+
+
 class Node:
-    def __init__(self, item, next):
+    def __init__(self, item: any, next):
```
Comment (dhruvmanila) [inline data_structures/linked_list/__init__.py:11]: There is no type hint called `any`. You probably meant `Any` which is imported from the `typing` module.

Also, please provide the type hint for the `next` parameter which probably will also be `Any` and return type hint as well which will be `None`.

File: data_structures/linked_list/__init__.py [hunk 5]
```diff
@@ -1,19 +1,28 @@
...
 
 
 class LinkedList:
-    def __init__(self):
+    def __init__(self) -> None:
         self.head = None
         self.size = 0
 
-    def add(self, item):
+    def add(self, item: any) -> None:
```
Comment (dhruvmanila) [inline data_structures/linked_list/__init__.py:21]:
```suggestion
    def add(self, item: Any) -> None:
```

File: data_structures/linked_list/__init__.py [hunk 6]
```diff
@@ -1,19 +1,28 @@
...
         self.head = Node(item, self.head)
         self.size += 1
 
-    def remove(self):
+    def remove(self) -> any:
```
Comment (dhruvmanila) [inline data_structures/linked_list/__init__.py:25]:
```suggestion
    def remove(self) -> Any:
```

File: data_structures/linked_list/__init__.py [hunk 7]
```diff
@@ -22,10 +31,33 @@ def remove(self):
...
+        """
+        >>> linked_list = LinkedList()
+        >>> linked_list.add(23)
+        >>> linked_list.add(14)
+        >>> linked_list.add(9)
+        >>> print(linked_list)
+        9 --> 14 --> 23
+        """
+        if not self.is_empty:
+            return "Linked List is empty."
```
Comment (dhruvmanila) [inline data_structures/linked_list/__init__.py:47]:
```suggestion
            return ""
```
This should be an empty string which suggests that the list is empty.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

File: data_structures/linked_list/__init__.py (hunk 7 above)
Comment (xcodz-dot) [reply data_structures/linked_list/__init__.py:47]: That's great (Although it might not be a good idea for production code but this one is educational).

**Comment by xcodz-dot** [comment]
try this solution every time pre-commit fails:
* `pip install pre-commit`
* `pre-commit run -a` (It might show you it failed but it actually did not)
* Commit the changes and push them into your fork

**Comment by arif599** [comment]
> 
> 
> try this solution every time pre-commit fails:
> 
>     * `pip install pre-commit`
> 
>     * `pre-commit run -a` (It might show you it failed but it actually did not)
> 
>     * Commit the changes and push them into your fork

Thanks for letting me know! I have done what you suggested. I'll commit again.

**Comment by dhruvmanila** [comment]
`pre-commit` is a hook you install in your local git directory. Once it is installed, it will run automatically every time you commit your changes, but this is only for the ***local*** directory.

If you've downloaded `pre-commit` locally, then go to the root of your Python repository copy and run `pre-commit install`. That's it! Now, you don't have to worry about the `pre-commit` checks ever failing.

**Comment by dhruvmanila** [comment]
@xcodz-dot If you give instructions to users about installing `pre-commit`, then please tell them to do `pre-commit install` instead of `pre-commit run -a`. This will save you and the user a ton of time and energy.

**Review by dhruvmanila — APPROVED** [review]
LGTM!

**Comment by arif599** [comment]
Seriously thank you guys for helping and guiding me. I'm a little new to this but I'll continue to learn and make more meaningful contributions along the way!

**Comment by xcodz-dot** [comment]
> @xcodz-dot If you give instructions to users about installing `pre-commit`, then please tell them to do `pre-commit install` instead of `pre-commit run -a`. This will save you and the user a ton of time and energy.

I will sure do that, i am just going to add that to contributing.md and will make a PR quickly

**Pull request merged**